File: 알고리즘/Algorithm/Algorithm/merge.cpp
#include <iostream>

using namespace std;
int a[10] = { 27,10,12,20,25,13,15,22,8,2 };

void m_sort(int list[], int l, int m, int r) {
	int i, j, k;
	i = l;
	j = m + 1;
	k = l;
	int temp[11];

	while (i <= m && j <= r) {
		if (list[i] <= list[j])
			temp[k++] = list[i++];
		else
			temp[k++] = list[j++];
	}

	if (i > m) {
		for (int x = j; x <= r; x++)
			temp[k++] = list[x];
	}
	else {
		for (int x = i; x <= m; x++)
			temp[k++] = list[x];
	}

	for (int x = l; x <= r; x++)
		list[x] = temp[x];
}

void merge(int list[], int l, int r) {
	int mid;
	if (l < r) {
		mid = (l + r) / 2;
		merge(list, l, mid);
		merge(list, mid + 1, r);
		m_sort(list, l, mid, r);
	}
}
int main() {
	merge(a, 0, 9);
	for (int i = 0; i < 10; i++) {
		cout << a[i] << " ";
	}
	return 0;
}

File: 알고리즘/Algorithm_JAVA/src/Merge.java
import java.util.Arrays;

public class Merge {
    public static int[] buff;

    public void merge(int[] data, int m, int mid, int n) {
        int i = m;
        int j = mid + 1;
        int k = m;

        while(i <= mid && j <= n) {
            if(data[i] <= data[j]) {
                buff[k] = data[i++];
            } else {
                buff[k] = data[j++];
            }
            k++;
        }

        if(i > mid) {
            for(int p = j; p <= n; p++, k++) {
                buff[k] = data[p];
            }
        } else {
            for(int p = i; p <- mid; p++, k++) {
                buff[k] = data[p];
            }
        }

        for(int p = m; p <= n; p++) {
            data[p] = buff[p];
        }
    }

    public void print(int[] data) {
        for(int i=0; i < data.length; i++) {
            System.out.print(data[i] + " ");
        }
    }

    public void mergeSort(int[] data, int m, int n) {
        int mid;
        if(m < n) {
            mid = (m+n)/2;
            mergeSort(data, m, mid);
            mergeSort(data, mid+1, n);
            merge(data, m, mid, n);
        }
    }
}
